### ISSUE TITLE
Look at making kernel-based images easier to extend and/or create

### ISSUE THREAD

**Issue by kevin-bates** [jupyter/enterprise_gateway#526]
This text is copied from #523 on behalf of @rayh...

I think it would be good to find a way to separate these kernel images from the main repo - would make it easier for people to contribute new ones as well as being able to select which ones you provide in your setup. I'm not sure how this would work, exactly, but it seems like every image requires a copy of kernel-launchers - perhaps that could be distributed as a tgz that anyone can pull into their own image? There are probably a few other constraints to how a kernel image needs to be setup

I'm also wondering how enterprise_gateway will know which kernels to use - could the whitelist be replaced with an explicit list of kernelspecs to use? perhaps as a JSON/YAML in s3 or suchlike? allowing alternatives/extras to be added. I think there are parts of the kernelspec that are internal (how to launch the kernel) and some metadata parts which I may like to override (such as display name, icon image, etc)

Perhaps the kernel image could accept kernel_id and response_address as env variables and is expected to start it's kernel by default (when no args are provided)? This would allow the kernel spec to be more focused on the metadata and delegate the responsibility of how to start a kernel into the kernel image itself

I think things like this would foster ownership and customisation of kernels by removing the need to understand the whole build process for enterprise_gateway (which, admittedly, is preventing me from doing a PR for the R+Spark image)

**Comment by kevin-bates** [comment]
@rayh - thanks for raising this issue!

>I'm not sure how this would work, exactly, but it seems like every image requires a copy of kernel-launchers - perhaps that could be distributed as a tgz that anyone can pull into their own image? There are probably a few other constraints to how a kernel image needs to be setup

I have thought about this very same thing - at least relative to packaging up the "glue" that allows for anyone to create their own kernel image.  I think you'd need the bootstrap-kernel.sh script along with the kernel-launcher relative to that language.

>I'm also wondering how enterprise_gateway will know which kernels to use - could the whitelist be replaced with an explicit list of kernelspecs to use? perhaps as a JSON/YAML in s3 or suchlike? 

EG locates available kernels exactly like all other jupyter apps, and I don't think we should stray from that model.  The whitelist used in the yaml file is purely an example - as is most of the settings in that file and works like a filter on what was found.  The idea would be that anyone adding kernels would create an applicable kernel.json file where the image is specified in the config section of the process-proxy stanza and update (or remove) the whitelist entry in the EG yaml file  That should work today.

>I think there are parts of the kernelspec that are internal (how to launch the kernel) and some metadata parts which I may like to override (such as display name, icon image, etc)

The kernelspecs EG provide are **purely** for example purposes.  Yes, they do the trick for most people, but, you're right, people need to turn these for their own application environments, etc.  That's perfectly acceptable.

>Perhaps the kernel image could accept kernel_id and response_address as env variables and is expected to start its kernel by default (when no args are provided)? 

I agree.  The kernel-image parameters are already set into env, but today we assume that they get repackaged as command-line arguments via the image's bootstrap script.  We could, and probably should, change each of the kernel launchers to first look at command line, then, if not present, check the env.  I had thought about that as well in my last pass.  

Where things get a little different with the k8s and docker kernel launch, is that EG launches an intermediate script - which, for k8s, reads in and processes the kernel-pod.yaml entries.   Much this model is purely to fit into the existing jupyter framework's model that is based on launching a sub-process, so we use the argv vector to specify what to run.  The other legacy piece is the replacement of the curly-bracketed values in the argv section with runtime information (e.g., {response_addr}) . Only this section gets substitutions performed - again based on a framework thing. 

Also remember that these kernel images cannot be run stand-alone (and expect things like interrupts to work).  If no response address is provided, they'd essentially be dead in the water.

So I'm thinking about this approach.
1. We document how to bring your own image.
2. We produce bundles that can be easily downloaded from our release (or wherever) and installed via Dockerfile - containing the bootstrap-kernel.sh and kernel-launcher files.
3. We also produce a template kernel.json file and contains the require substitution parameters and argv vector.  This could be part of the documentation of item 1.

On a personal note...I am on vacation until January 7, 2019 as is another primary contributor.  That said, I will be spending a few hours each week to make sure no one is blocked and keeping issues active as best as possible.  Just wanted to let you know that.

**Comment by rayh** [comment]
What is the recommended way to update the kernelspecs in EG?  Could I use some sort of configmap in k8s?

**Comment by kevin-bates** [comment]
We recommend that the kernelspecs be loaded into a PV that is mounted in the EG pod.  In fact, the `enterprise-gateway.yaml` file has commented out statements (relative to NFS) to do just that.

Are you referring to things like `spark-context-initialization-mode`?  I suppose that particular argument could be removed from the non-spark-based kernelspecs.  I'll make a note to do that when I package up the kernel-launchers into a distribution that can then be used from a Dockerfile.

**Comment by kevin-bates** [comment]
@rayh - I meant to comment on this statement from the opening section...

>[...]and some metadata parts which I may like to override (such as display name, icon image, etc)

i just wanted you to know that pieces of the kernelspec that are considered _resources_ are not supported by either of the gateway projects.  An example would be the icon files to display in the notebook when the kernel is running.  The kernelspec resources currently make an assumption that they're available for upload by the browser (i.e., they reside on the notebook server itself).  However, the NB2KG and Gateway implementations do not support the ability for those _resources_ to be pulled from remote gateway servers.

**Comment by rayh** [comment]
Ok, I've created my own kernels: https://github.com/eliiza/jupyter-kernels (and built images are here: https://cloud.docker.com/u/eliiza/repository/list).  I changed the base spark image to extend from base-notebook so it had some of the basic assumptions that the other kernels do (OS version, etc), mainly to make it easy for the user.   The kernelspecs are uploaded into the EG PV

Right now, I just have the kernel-launcher scripts hard-coded into that project but would like to pull the latest from this project - let me know if/when you have a mechanism to do that.

This seems to be working well so far.
